validation/deep-compare.js

```diff
@@ -0,0 +1,194 @@
+      const deleteCost = previousRow[j] + 1;
+      const replaceCost = previousRow[j - 1] + (strOne[i - 1] !== strTwo[j - 1] ? 1 : 0);
+      currentRow[j] = Math.min(insertCost, deleteCost, replaceCost);
+    }
+    previousRow = currentRow;
+  }
+
+  return previousRow[strTwo.length];
+}
+
+/**
+ * Calculates the similarity between two strings using the Levenshtein distance algorithm.
+ *
+ * @param {string} strOne - The first string.
+ * @param {string} strTwo - The second string.
+ * @returns {number} The similarity score between the two strings. (0.0 to 1.0)
+ */
+export function lSimilarity(strOne, strTwo) {
+  const distance = lDistance(strOne, strTwo);
+  const maxLength = Math.max(strOne.length, strTwo.length);
+  return 1 - distance / maxLength;
+}
+
+/**
+ * Converts a similarity percentage into a readable category.
+ *
+ * @param {number} similarity - The similarity score between the two strings. (0.0 to 1.0)
+ * @returns {string} The readable similarity category.
+ */
+export function similarityCategory(similarity) {
+  const categories = [
+    { range: [0, 0.2], value: SIMILARITY_NONE },
+    { range: [0.2, 0.5], value: SIMILARITY_LOW },
+    { range: [0.5, 0.7], value: SIMILARITY_MEDIUM },
+    { range: [0.7, 1.0], value: SIMILARITY_HIGH },
+    { range: [1.0, 1.1], value: SIMILARITY_EXACT },
+  ];
+  const findCategory = (option) => option.range[0] <= similarity && option.range[1] > similarity;
+  const category = categories.find(findCategory);
+
+  return category?.value || SIMILARITY_UNKNOWN;
+}
+
+/**
+ * Compares two strings and returns an object containing observations about their similarity.
+ *
+ * @param {string} stringOne - The first string to compare.
+ * @param {string} stringTwo - The second string to compare.
+ * @returns {Object} - An object containing observations about the similarity of the two strings.
+ */
+export function observe(stringOne = '', stringTwo = '') {
+  const observations = {};
+
+  observations[MATCH] = stringOne === stringTwo;
+  observations[EMPTY] = (!stringOne || !stringTwo);
+  observations[BOTH_EMPTY] = (!stringOne && !stringTwo);
+  observations[LENGTHS_MATCH] = stringOne.length === stringTwo.length;
+
+  const similarity = lSimilarity(stringOne, stringTwo);
+
+  observations[SIMILARITY_PERCENTAGE] = `${Math.round(similarity * 100)}%`;
+  observations[SIMILARITY] = similarityCategory(similarity);
+  observations[WHITESPACE] = (stringOne.replace(/\s/g, '') === stringTwo.replace(/\s/g, ''));
+  observations[ASCII] = [stringOne, stringTwo].filter((string) => string.match(/%20/g)).length > 0;
+
+  return observations;
+}
+
+/**
+ * Observes the changes between two text values.
+ *
+ * @param {string} oldText - The old text value.
+ * @param {string} newText - The new text value.
+ * @returns {Object} - The observed changes between the old and new text values.
+ */
+export function observeText(oldText = '', newText = '') {
+  return observe(oldText, newText);
+}
```

Comment:
I was expecting to have different observations for URLs and for Texts but all the text ones are shared in `observe`.
We can omit this function and bring it back if/when that becomes true.

Reply:
Actually I think it makes more sense to rename `observe` to `observeText` and have `observeUrl` call `observeText` instead of the generic `observe` function.